[소나무] Step-1 회원가입, 회원목록 조회 기능 부분 구현

daewon-ko opened the pull request:
##  PR 작성 전 체크 리스트

안녕하세요. 이번에 리뷰를 받게 된 소나무라고 합니다.
자바 언어도 잘 다루지 못하는 상황이고, 스프링 구현은 커녕 학습조차 생소해서 모든게 생경하였습니다. 
이번주 환절기 감기와 컨디션 저하로 인해서 미션을 충실히 진행하지 못했습니다. 
첫인상이 중요한 법인데 죄송합니다. 

커밋 단위도 초반에는 정신이 없어서 한 번에 올리다가 뒷부분부터 조금씩 나누려고 노력했습니다. 
초반 리드미 학습계획에는 이런저런 강의를 듣고 구현을 시도하려고 했지만 CS기간 때부터
학습만 하다보니 구현을 시도조차 못하는 경우가 더러 존재해서 이번에는 잘 모를지라도 일단 
돌아가는 것을 목표로 프로그래밍을 했습니다. 

다소 부족한 점이 많이 존재하겠지만 너그러이 봐주시면 참으로 감사하겠습니다.
시간 관계상 일단 한 곳까지만 올리도록 하겠고, 주말 등을 이용하여 추가해서 계속 보강하겠습니다. 
그리고 이번주는 다소 실망스러우실 수 있지만, 다음주부터는 이번주보단  조금 더 노력하도록 하겠습니다.

---
### 학습키워드
* HTTP GET, POST
* GetMapping, PostMapping
* Controller, view, Model
* Model Attribute
* Handler
* 타임리프 문법
---
### 주요기능
[x] 회원가입 기능 구현
[x] 회원 전체 조회 기능 구현
[ ] 회원 프로필 조회 기능 구현
[ ] 회원 프로필 수정 기능 구현
[]UserService 클래스 테스트코드 작성

### 고민한부분
*HTML 학습도 온전치 못한 상황에서 타임리프 등을 적용하는데 상당히 애를 먹었습니다.
* 클래스 설계를 잘 못하지만, DTO 등은 현재단위에선 필요없을 것이라 생각하여 제외하였습니다.
(DTO, DAO 등에 부분에 대해서도 추후 학습이 필요합니다.)
* 금일 스쿼드세션 중 저의 예외처리 부분에 대해서 실제로 터트려봤는데 작동하지 않았습니다. 이부분은 추후 보강이 필요할 것 같습니다

daewon-ko committed:
feat "README 작성"

학습계획을 간단하게 작성하였습니다.

daewon-ko committed:
feat "static/qna 내 html 파일 template 폴더로 이동"

templates/qna/form.html의 간단한 타임리프 문법 추가
(원인은 알 수 없지만, form 형식 등에서 css 지원 불가.)

daewon-ko committed:
feat "User, UserRepository, UserService 생성"

1. User 클래스 생성(Id, password, email, name의 필드를 갖고 나머지는 getter,setter로 전달)
2. UserRepository 클래스 생성(User 클래스에 의존하고 HashMap을 이용. save, findByEmail, findAllUser 메서드 생성)
3. UserService 클래스 생성(UserRepository, User 클래스에 의존. UserRepository를 필드로 두고 생성자 주입 이용.
또한 User 클래스를 활용하여 비즈니스 로직(join)인 메서드를 생성

daewon-ko committed:
refactor "UserController 내 GetMapping, PostMapping 메서드 수정 및 template 내 HTML 경로 변경"

GetMapping 및 PostMapping 적용오류 현상 발생.
페어리뷰 중 도움을 받아서 메서드의 return 값등 변경.

daewon-ko committed:
refactor "list.html 파일 href 경로 수정"

타임리프 문법 적용 후 css파일 적용을 위해 link를 수정

daewon-ko committed:
refactor "form.html 내 a 태그 경로 변경"

a 태그내 경로 타임리프 문법 적용 및 경로설정 변경

daewon-ko committed:
feat "UserService 내 showAllUser 메서드 추가"

Controller 내 GetMapping 메서드에서 전체 회원목록 조회를 위한 목적의 메서드 추가

daewon-ko committed:
feat "UserController 내 getUserList 메서드 추가"

daewon-ko committed:
refactor "UserController register(PostMapping 메서드) Return 형식 문제 조건과 같이 변경"

daewon-ko committed:
refactor "1-2 회원목록 기능구현 중 구현 이상으로 인해 수정"

타임리프 문법 적용 중 조원들의 도움을 받아 th:text : user.getId() -> user.id와 같은 형식으로 변경
form.html 파일 중 name="userID"(기본값) -> name="id"(User 클래스 내 멤버변수명과 동일하게 설정)로 변경
이후 정상적으로 기능작동 완료

daewon-ko committed:
refactor "UserController 내 필요없는 주석 제거"

daewon-ko committed:
refactor "스쿼드 세션 후 refactoring"

1. UserService 내 checkDuplicatedId 메서드가 예외를 실제로 발생하는 것을 확인하고 싶었으나 작동하지 않았습니다.
2. 타고 들어가서 UserRepository 내 findByEmail 메서드의 리턴타입 등이 문제가 있는 것으로 추정되어서 email 값을 리턴하도록 수정하였습니다.
3. 아직은 정확한 원인은 알 수 없지만, 추후 원인을 찾아보겠습니다.

sallyjellyy commented on src/main/java/kr/codesqaud/cafe/Repository/UserRepository.java:
```diff
@@ -0,0 +1,29 @@
+
+    public Optional<String> findByEmail(String id) {
+        return Optional.ofNullable(store.get(id).getEmail());
+    }
```
메서드 이름과 파라미터가 잘못 된 것 같습니다. 에러를 의도적으로 터뜨려보셨다고 하셨는데 이 부분과 연관이 있어보이네요. 메서드명, 파라미터, 리턴하는 값을 체크해보시길 바랍니다.

sallyjellyy commented on src/main/java/kr/codesqaud/cafe/Repository/UserRepository.java:
```diff
@@ -0,0 +1,29 @@
+
+    public void clearStore() {
+        store.clear();
+    }
```
사용되지 않는 코드로 보입니다. 필요시에만 작성하실 것을 권장합니다.

sallyjellyy commented on src/main/java/kr/codesqaud/cafe/controller/UserController.java:
```diff
@@ -0,0 +1,43 @@
+
+    private final UserService userService;
+
+    @Autowired
```
`@Autowired` 는 생략 가능합니다. 의존성 주입에 관해 알아보시면 좋을 것 같아요.

sallyjellyy commented on src/main/java/kr/codesqaud/cafe/domain/User.java:
일부 getter, 모든 setter 가 사용되지 않는걸로 보이네요. 여기서도 필요한 메서드만 작성하시는 것을 권장합니다.

sallyjellyy commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -0,0 +1,35 @@
+        this.userRepository = userRepository;
+    }
+
+    public void join(User user) {
+        checkDuplicateId(user.getEmail());
+        userRepository.save(user);
+    }
+
+    private void checkDuplicateId(String email) {
+        userRepository.findByEmail(email)
+                .ifPresent(m -> {
+                    throw new IllegalArgumentException("이미 존재하는 회원입니다.");
+                });
+    }
```
에러를 일부러 터뜨리신 부분 같네요. 해당 에러는 `UserRepository`에 남긴 리뷰를 반영하시면 의도하신대로 작동할 것 같습니다.

에러와 별개로 `checkDuplicatedId` 는 메서드명으로 추론했을 때 같은 Id를 가진 유저가 있는지를 결과값으로 리턴할 것 같아요.
`join` 에서 중복되는 아이디의 유저가 있는지 확인하고 있다면 거기서 예외를 던지는게 더 자연스럽지 않나 생각됩니다. 물론 의도대로 작동한다면 현재 구현도 좋습니다.

sallyjellyy commented on src/main/java/kr/codesqaud/cafe/service/UserService.java:
```diff
@@ -0,0 +1,35 @@
+public class UserService {
+    private final UserRepository userRepository;
+
+    @Autowired
```
여기도 마찬가지예요!

sallyjellyy reviewed (changes requested):
안녕하세요, 소나무! 새리입니다.

스프링 처음이라고 하셨는데 고생 많이 하셨을 것 같네요. 개인적으로는 미션을 얼마나 진행했는지는 크게 중요하지 않고 미션이 의도하는 학습거리에 대해 제대로 공부만 할 수 있다면 그걸로 족하다고 생각합니다. 그러니 너무 조급해하지 않으셔도 됩니다. 천천히 건강에 무리가지 않는 선에서 공부하시고 미션에 녹여내시면 될 것 같아요.

학습 계획 쓰신걸 봤는데 하나 권장드리는 점은(이미 하고 계실 수도 있지만!) 하루하루 하신 것(학습 내용, 느낀점 등)을 짧게 요약해서 같이 정리하시면 스스로 트랙킹이 수월해지실거예요.

앞으로도 화이팅입니다.

daewon-ko committed:
refactor "UserService 내 메서드 수정"

checkDuplicated 메서드의 매개변수로 User 객체의 id로 둠.
findByEmail의 리턴값이 email(String형)이므로 해당 email으로 중복검사하여
값이 중복될경우 Exception처리 하도록 설계

daewon-ko committed:
refactor "필요없는 메서드 제거"

clearStore 제거

daewon-ko committed:
refactor "리뷰 기반 리팩토링 진행"

1. findByEmail 메서드명에따라 매개변수 변경 및 stream 연산 이용하여 메서드 내부 로직 변경
2. User 객체 필요한 getter 메서드 제외하고 삭제.
:

daewon-ko committed:
refactor "불필요 주석 삭제"

sallyjellyy commented:
고생하셨습니다~!
